source/common/upstream/cluster_manager_impl.cc

```diff
@@ -40,6 +40,16 @@
 
 namespace Envoy {
 namespace Upstream {
+namespace {
+
+void addOptionsIfNotNull(Network::Socket::OptionsSharedPtr& options,
+                         const Network::Socket::OptionsSharedPtr& to_add) {
+  if (to_add != nullptr) {
```

Comment:
Can this ever happen? Maybe just ASSERT?

Reply:
The code that I refactored it away from had it as possible. Digging in a little deeper, the socketOptions call on `Network::Connection` returns a shared_ptr reference, but it's not clear from the API  that it is required to be non-null.

`Network::ConnectionImpl` just returns the shared_ptr from `Network::SocketImpl`

Looking at `Network::SocketImpl`, it only sets the options if an option has previously been added. I'm a little uncomfortable assuming that it won't be null because of that.

I really wish C++ had a better way of sharing ownership of an object while requiring that it is set to something valid. :/

Reply:
OK fine to leave. I think you could make a shared_ptr override that does not allow nullptr values, but that's a different topic. :)